### PR TITLE
Enhancement on applying themed icon

### PR DESCRIPTION
**Pull request by naveenccmsd**
## Description

Working on some feature enhancement and fix

New Features
- Apply themed icons for system icons if they support (android 13+ [not sure about android 12] )
- Select Icon pack and Themed Icon separately for home and app drawer (This help to apply Themed icons (home) + delta icons (app drawer) together) - Ref: https://github.com/LawnchairLauncher/lawnicons/pull/815#issuecomment-1342004064
- Support Transparant background on adaptive icons 

Limitation Identified :

- When themed enabled for home page alone,  only icons that are from selected icons pack will be themed , custom icon selected manual will not themed (including lawnicons). No issue if both home and drawer is selected.

 ~~Themed icons are can't applied to home screen alone from IconPacks. Below change would disable _Home Screen_ option for iconpacks and apply themed icons from lawnincons by default if installed.~~

- ~~When system icons are selected - Both _Home Screen_ and **Home Screen & App Drawer option** will be enabled and lawnicons are applied by default~~
- ~~when Themed support icon packs are selected (including lawnicons) - Only  **Home Screen & App Drawer option** enabled~~ 
- ~~When Other Icons packs slected which don't support thhemed icons - **Not Supported.** option will be displayed~~

:white_check_mark:Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

File: lawnchair/src/app/lawnchair/ui/preferences/IconPackPreferences.kt [hunk 1]
```diff
@@ -159,6 +165,8 @@ fun IconPackPreferences() {
                     },
                     description = if (themedIconsAvailable.not()) {
                         stringResource(id = R.string.lawnicons_not_installed_description)
+                    } else if(!themedIconPack.contains(iconPackAdapter.state.value) && iconPackAdapter.state.value != "") {
```
Comment (SuperDragonXD) [inline lawnchair/src/app/lawnchair/ui/preferences/IconPackPreferences.kt]: Move the `if (...)` contents to a variable (like `iconPackSupportsThemedIcon`, `supportsThemedIconPack`, or something else), so you don't have to repeat this code for the above comment:

```kotlin
// somewhere near the top
val supportsThemedIconPack = themedIconPack.contains(iconPackAdapter.state.value) && iconPackAdapter.state.value != ""

// the code in here:
                    } else if(!supportsThemedIconPack) {
```

File: lawnchair/src/app/lawnchair/ui/preferences/IconPackPreferences.kt [hunk 2]
```diff
@@ -135,15 +135,21 @@ fun IconPackPreferences() {
...
             PreferenceGroup {
-                val themedIconsAvailable = LocalContext.current.packageManager
-                    .getThemedIconPacksInstalled(LocalContext.current)
+                val themedIconPack = LocalContext.current.packageManager.getThemedIconPacksInstalled(LocalContext.current)
+                val themedIconsAvailable = themedIconPack
                     .any { LocalContext.current.packageManager.isPackageInstalled(it) }
                     || LocalContext.current.packageManager
                     .isPackageInstalled(Constants.LAWNICONS_PACKAGE_NAME)
                 ListPreference(
                     enabled = themedIconsAvailable,
```
Comment (SuperDragonXD) [inline lawnchair/src/app/lawnchair/ui/preferences/IconPackPreferences.kt]: I'd suggest disabling the preference when the icon pack doesn't support themed icons:
```suggestion
                    enabled = themedIconsAvailable && supportsThemedIconPack,
```

File: lawnchair/res/values/strings.xml [hunk 3]
```diff
@@ -124,6 +124,7 @@
       <string name="themed_icons_home_label">Home Screen</string>
       <string name="themed_icons_home_and_drawer_label">Home Screen &amp; App Drawer</string>
       <string name="lawnicons_not_installed_description">Lawnicons not installed.</string>
+      <string name="themed_icons_not_supported">Not Supported.</string>
```
Comment (SuperDragonXD) [inline lawnchair/res/values/strings.xml]:
```suggestion
      <string name="themed_icons_not_supported">Unsupported by Icon Pack.</string>
```

**Review by SuperDragonXD — CHANGES_REQUESTED** [review]
Thanks for the PR! Have a few UI suggestions in mind.

(Also if you have the time, I'd recommend following my comment on https://github.com/LawnchairLauncher/lawnicons/pull/815#issuecomment-1342004064 (just the "UI for applying Lawnicons") to make the UI more clearer, and so you can apply X icon pack normally and Y icon pack for themed icons)

File: lawnchair/src/app/lawnchair/ui/preferences/IconPackPreferences.kt (hunk 1 above)
Comment (naveenccmsd) [reply lawnchair/src/app/lawnchair/ui/preferences/IconPackPreferences.kt]: Can't use  both themedIconPack.contains(iconPackAdapter.state.value) && iconPackAdapter.state.value != "" to variable as usuage is different but I have created another variable for better readable as you suggested.

**Review by SuperDragonXD — APPROVED** [review]

**Comment by naveenccmsd** [comment]
> Thanks for the PR! Have a few UI suggestions in mind.
> 
> (Also if you have the time, I'd recommend following my comment on [LawnchairLauncher/lawnicons#815 (comment)](https://github.com/LawnchairLauncher/lawnicons/pull/815#issuecomment-1342004064) (just the "UI for applying Lawnicons") to make the UI more clearer, and so you can apply X icon pack normally and Y icon pack for themed icons)

Yes , I will try to change UI as future request.

**Comment by suphon-t** [comment]
I'm confused with this. What if I want to use Delta everywhere else, and have themed icons on the home screen? Is that impossible?

**Comment by SuperDragonXD** [comment]
Yeah, that's currently impossible on the current UI. (See my above comment)

**Comment by naveenccmsd** [comment]
> I'm confused with this. What if I want to use Delta everywhere else, and have themed icons on the home screen? Is that impossible?

Yes correct, can't apply themed icons home drawer alone when themed applied from icons pack, that's why update code to apply lawnicons theme if system icons are selected.

To work themed icons along with delta required more changes in how themed icons are applied along with ui changes suggested by @SuperDragonXD ..


May be I can mark this pr as draft and work on code those changes. As this ui needs to be updated any way.

File: lawnchair/src/app/lawnchair/ui/preferences/IconPackPreferences.kt [hunk 4]
```diff
@@ -156,12 +155,48 @@ fun IconPackPreferences() {
...
+                        iconPackAdapter.onChange(newValue = "")
+                        themedIconPackAdapter.onChange(newValue = "")
                     },
                     description = if (themedIconsAvailable.not()) {
                         stringResource(id = R.string.lawnicons_not_installed_description)
                     } else null,
                 )
             }
+
+            ExpandAndShrink(visible = themedIconsAdapter.state.value && !drawerThemedIconsAdapter.state.value) {
```
Comment (SuperDragonXD) [inline lawnchair/src/app/lawnchair/ui/preferences/IconPackPreferences.kt:167]: How about you move `drawerThemedIconsAdapter.state.value` to a new variable, like `drawerThemedIconsEnabled`?

Comment (SuperDragonXD) [inline lawnchair/src/app/lawnchair/ui/preferences/IconPackPreferences.kt]: Move `drawerThemedIconsAdapter.state.value` to a new variable, like `drawerThemedIconsEnabled`, to shorten the variable name.

File: lawnchair/src/app/lawnchair/ui/preferences/IconPackPreferences.kt [hunk 5]
```diff
@@ -156,12 +155,48 @@ fun IconPackPreferences() {
...
+                PreferenceGroup {
+                    PreferenceGroupHeading(heading = stringResource(id = R.string.themed_icon_pack))
```
Comment (SuperDragonXD) [inline lawnchair/src/app/lawnchair/ui/preferences/IconPackPreferences.kt]:
```suggestion
                PreferenceGroup (
                    heading = stringResource(id = R.string.themed_icon_pack)
                ) {
```

The same should be applied to other headings.

File: lawnchair/src/app/lawnchair/ui/preferences/IconPackPreferences.kt [hunk 6]
```diff
@@ -156,12 +155,48 @@ fun IconPackPreferences() {
...
+                    IconPackGrid(
+                        adapter = themedIconPackAdapter,
+                        modifier = Modifier.padding(bottom = 6.dp, top = 6.dp),
```
Comment (SuperDragonXD) [inline lawnchair/src/app/lawnchair/ui/preferences/IconPackPreferences.kt]: Maybe move this to IconPackGrid, so it's DRY.

File: lawnchair/src/app/lawnchair/ui/preferences/IconPackPreferences.kt [hunk 7]
```diff
@@ -170,14 +205,32 @@ fun IconPackPreferences() {
...
+    val themedIconPacksName = themedIconPacks.map { it.name }
+
+    if (isThemedIconPack) {
+        _iconPacks = if (drawerThemedIcons) {
+            themedIconPacks
+        } else {
+            themedIconPacks.filter { it.packageName != "" }
+        }
+    } else {
+        if (drawerThemedIcons) {
```
Comment (SuperDragonXD) [inline lawnchair/src/app/lawnchair/ui/preferences/IconPackPreferences.kt]: Move this to an `else if`.

**Review by SuperDragonXD — CHANGES_REQUESTED** [review]
Thanks for implementing the suggested changes! While the current UI is much more workable, it isn't exactly what I meant though.

I'd suggest moving all the themed icon preferences to a new file, `ThemedIconPreferences.kt`. Therefore, the UI will look much more polished and less cramped. It also makes code maintenance easier. (Should have made this intention much clearer in my original comment.)

If that's not possible, I do have a few changes in mind for the original UI.

**Comment by naveenccmsd** [comment]
@SuperDragonXD I have applied suggested changes.

File: lawnchair/res/values/strings.xml [hunk 8]
```diff
@@ -117,6 +117,7 @@
     <!-- IconPackPreferences -->
     <!-- <string name="icon_style" /> -->
     <string name="icon_pack">Icon Pack</string>
+    <string name="themed_icon_pack">Themed Icon Pack</string>
```
Comment (SuperDragonXD) [inline lawnchair/res/values/strings.xml]:
```suggestion
    <string name="themed_icon_pack">Themed Icon Source</string>
```
"Themed Icon Pack" seems vague, so this.

**Review by SuperDragonXD — CHANGES_REQUESTED** [review]
Thanks for applying the suggested changes, and sorry for taking so long to review.

Anyways, there seems to be a bug when choosing the "Home Screen and App Drawer" option (and switching back, not shown):

https://user-images.githubusercontent.com/70206496/211120851-ec784ad4-aa0a-4116-a990-fd6cda0241e4.mp4

Also, I'd suggest auto-selecting Lawnicons in the UI when switching to the "Home Screen" option.

Lastly, maybe move the "Icon Pack" option above the "Themed Icons" dropdown like before? (although it will take some additional changes) So it makes more sense when reading top to bottom.

Example pseduocode:
```
ExpandAndShrink {
  IconPackSelector()
}

PreferenceGroup {
  ListPreference ("Themed Icons")
  ExpandAndShrink {
    ThemedIconSourceSelector()
  }
  ExpandAndShrink {
    DrawerThemedIconSourceSelector()
  }
}
```
(If you do that, the preferencegroupheaders should be similar to normal label text)

**Comment by naveenccmsd** [comment]
> (If you do that, the preferencegroupheaders should be similar to normal label text)

I have made some changes,  Now the iconpack would be automatically selected if you switch between _off_ & _Home & Drawer_ . But It needs to reset for _Home_ option so user can select both themed and normal icon pack. If I don't do that previous selected icons are fectched from cache.

**Review by SuperDragonXD — APPROVED** [review]
Looks good now. I'll ask some team members internally to see if they experience any other bugs, then I'll merge this PR

**Comment by naveenccmsd** [comment]
> Looks good now. I'll ask some team members internally to see if they experience any other bugs, then I'll merge this PR

Thanks

**Comment by SuperDragonXD** [comment]
I've discussed internally, and the bug that I said earlier still persists in the latest commit. Could you fix it first?

**Comment by SuperDragonXD** [comment]
cc @naveenccmsd see above comment (in case you haven't seen it yet)

**Comment by AR-Home** [comment]
https://katb.in/acisakekilo

**Comment by SuperDragonXD** [comment]
Since there's still no reply yet, I'll merge this PR and try to fix the issue myself.

**Comment by naveenccmsd** [comment]
> Since there's still no reply yet, I'll merge this PR and try to fix the issue myself.

Hey @SuperDragonXD , Sorry I was busy on some personal work, Thanks for merging the PR.

**Comment by SuperDragonXD** [comment]
It's fine. Although if you have a patch for the mentioned issue, feel free to open a PR for it. I think I might not have time to fix it myself, due to some personal things.

**Pull request merged**